File: library/src/main/java/com/carlosmuvi/segmentedprogressbar/SegmentedProgressBar.java
package com.carlosmuvi.segmentedprogressbar;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.support.annotation.ColorInt;
import android.util.AttributeSet;
import android.view.View;

/**
 * Created by carlosmuvi on 02/09/16.
 */

public class SegmentedProgressBar extends View {

    private int segmentCount = 5;
    private int lastCompletedSegment = 0;
    private int currentSegmentProgressInPx = 0;

    private Paint containerRectanglePaint;
    private Paint fillRectanglePaint;
    private DrawingTimer drawingTimer;

    public SegmentedProgressBar(Context context) {
        super(context);
        initView();
    }

    public SegmentedProgressBar(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView();
    }

    private void initView() {
        initDrawingTimer();
        containerRectanglePaint = buildContainerRectanglePaint(Color.LTGRAY);
        fillRectanglePaint = buildFillRectanglePaint(Color.RED);
    }

    private void initDrawingTimer() {
        drawingTimer = new DrawingTimer();
        drawingTimer.setListener(new DrawingTimer.Listener() {
            @Override public void onTick(int currentTicks, int totalTicks) {

                int segmentWidth = getSegmentWidth();

                currentSegmentProgressInPx = currentTicks * segmentWidth / totalTicks;
                if (totalTicks <= currentTicks) {
                    lastCompletedSegment++;
                    currentSegmentProgressInPx = 0;
                }
                invalidate();
            }
        });
    }

    @Override protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        drawContainerRectangles(canvas);
        drawCompletedRectangles(canvas);
        drawCurrentRectangle(canvas);
    }

    /*
    EXPOSED ATTRIBUTE METHODS
     */

    public void setContainerColor(@ColorInt int color) {
        containerRectanglePaint = buildContainerRectanglePaint(color);
    }

    public void setFillColor(@ColorInt int color) {
        fillRectanglePaint = buildFillRectanglePaint(color);
    }

    public void setSegmentCount(int segmentCount) {
        this.segmentCount = segmentCount;
    }

    /*
    EXPOSED ACTION METHODS
     */

    public void playSegment(long timeInMilliseconds) {
        if (!drawingTimer.isRunning()) {
            drawingTimer.start(timeInMilliseconds);
        }
    }

    public void pause() {
        drawingTimer.pause();
    }

    public void setCompletedSegments(int completedSegments) {
        if (completedSegments <= segmentCount) {
            lastCompletedSegment = completedSegments;
            invalidate();
        }
    }

    /*
    PRIVATE METHODS
     */

    private void drawContainerRectangles(Canvas canvas) {
        int segmentWidth = getSegmentWidth();

        int leftX = 0;
        int rightX = leftX + segmentWidth;
        int topY = 0;
        int botY = getHeight();

        for (int i = 0; i < segmentCount; i++) {
            drawRoundedRect(canvas, leftX, topY, rightX, botY, containerRectanglePaint);
            leftX = leftX + segmentWidth + getSegmentGapWidth();
            rightX = leftX + segmentWidth;
        }
    }

    private void drawCompletedRectangles(Canvas canvas) {
        int segmentWidth = getSegmentWidth();

        int leftX = 0;
        int rightX = leftX + segmentWidth;
        int topY = 0;
        int botY = getHeight();

        for (int i = 0; i < lastCompletedSegment; i++) {
            drawRoundedRect(canvas, leftX, topY, rightX, botY, fillRectanglePaint);
            leftX = leftX + segmentWidth + getSegmentGapWidth();
            rightX = leftX + segmentWidth;
        }
    }

    private void drawCurrentRectangle(Canvas canvas) {
        int segmentWidth = getSegmentWidth();

        int leftX = lastCompletedSegment * (segmentWidth + getSegmentGapWidth());
        int rightX = leftX + currentSegmentProgressInPx;
        int topY = 0;
        int botY = getHeight();
        drawRoundedRect(canvas, leftX, topY, rightX, botY, fillRectanglePaint);
    }

    private void drawRoundedRect(Canvas canvas, float left, float top, float right, float bottom, Paint paint) {

        Path path = new Path();
        float rx = 6;
        if (rx < 0) rx = 0;
        float ry = 6;
        if (ry < 0) ry = 0;
        float width = right - left;
        float height = bottom - top;
        if (rx > width / 2) rx = width / 2;
        if (ry > height / 2) ry = height / 2;
        float widthMinusCorners = (width - (2 * rx));
        float heightMinusCorners = (height - (2 * ry));

        path.moveTo(right, top + ry);
        path.rQuadTo(0, -ry, -rx, -ry);//top-right corner
        path.rLineTo(-widthMinusCorners, 0);
        path.rQuadTo(-rx, 0, -rx, ry); //top-left corner
        path.rLineTo(0, heightMinusCorners);

        path.rQuadTo(0, ry, rx, ry);//bottom-left corner
        path.rLineTo(widthMinusCorners, 0);
        path.rQuadTo(rx, 0, rx, -ry); //bottom-right corner

        path.rLineTo(0, -heightMinusCorners);

        path.close();//Given close, last lineto can be removed.

        canvas.drawPath(path, paint);
    }

    private Paint buildFillRectanglePaint(@ColorInt int color) {
        Paint paint = new Paint();
        paint.setColor(color);
        paint.setStyle(Paint.Style.FILL);
        return paint;
    }

    private Paint buildContainerRectanglePaint(@ColorInt int color) {
        Paint paint = new Paint();
        paint.setColor(color);
        paint.setStyle(Paint.Style.FILL);
        return paint;
    }

    private int getSegmentWidth() {
        return (getWidth() / segmentCount) - getSegmentGapWidth();
    }

    private int getSegmentGapWidth() {
        return dpToPx(1);
    }

    private int dpToPx(int valueInDp) {
        float density = getContext().getResources().getDisplayMetrics().density;
        return (int) (valueInDp * density);
    }
}

File: library/src/main/java/com/carlosmuvi/segmentedprogressbar/SegmentedProgressBar.kt
package com.carlosmuvi.segmentedprogressbar

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.util.Log

import android.view.View
import androidx.annotation.ColorInt
import java.math.BigDecimal

class SegmentedProgressBar : View
{

    private var lastCompletedSegment = 0
    private var currentPartialCompletedSegment = 0f
    private var currentSegmentProgressInPx = 0

    private lateinit var containerRectanglePaint: Paint
    private lateinit var fillRectanglePaint: Paint
    private lateinit var drawingTimer: DrawingTimer
    private lateinit var properties: PropertiesModel

    private var segmentCompletedListener: CompletedSegmentListener? = null

    constructor(context: Context) : super(context)
    {
        initView()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    {
        initView(attrs)
    }

    private fun initView(attrs: AttributeSet? = null)
    {
        initDrawingTimer()
        initPropertiesModel(attrs)

        containerRectanglePaint = buildContainerRectanglePaint(properties.containerColor)

        fillRectanglePaint = buildFillRectanglePaint(properties.fillColor)

           }

    private fun initPropertiesModel(attrs: AttributeSet?)
    {
        properties = PropertiesModel(context, attrs)


        val sss = properties.progressFilledContainers.subtract(BigDecimal(properties.progressFilledContainers.toInt())).toFloat()

        setCompletedSegments(properties.progressFilledContainers.toInt(), sss)
    }

    private fun initDrawingTimer()
    {
        drawingTimer = DrawingTimer()



        drawingTimer.setListener { currentTicks, totalTicks ->
            val segmentWidth = segmentWidth



            currentSegmentProgressInPx = currentTicks * segmentWidth / totalTicks
            if (totalTicks <= currentTicks)
            {
                lastCompletedSegment++
                currentSegmentProgressInPx = 0
            }
            if (totalTicks == currentTicks)
            {
                if(currentPartialCompletedSegment!=0f)
                    currentPartialCompletedSegment=0f
                segmentCompletedListener?.onSegmentCompleted(lastCompletedSegment)
            }
            invalidate()
        }
    }

    override fun onDraw(canvas: Canvas)
    {
        super.onDraw(canvas)

        drawContainerRectangles(canvas)
        drawCompletedRectangles(canvas)
        drawCurrentRectangle(canvas)
    }

    /**
     * Set the color of the unfilled part of the progress bar.
     */
    @Suppress("Unused")
    fun setContainerColor(@ColorInt color: Int)
    {
        containerRectanglePaint = buildContainerRectanglePaint(color)
    }

    /**
     * Set the color of the filled part of the progress bar.
     */
    @Suppress("Unused")
    fun setFillColor(@ColorInt color: Int)
    {
        fillRectanglePaint = buildFillRectanglePaint(color)
    }

    var containerColor: HashMap<Int, Int> = HashMap<Int, Int>()

    fun setFillColor(@ColorInt color: Int, segmentIndex: Int)
    {
        containerColor.put(segmentIndex, color)
        properties.segmentedContainerColors = containerColor
    }

    /**
     * Set the total number of segments that compose the progress bar.
     */
    @Suppress("Unused")
    fun setSegmentCount(segmentCount: Int)
    {
        properties.segmentCount = segmentCount
    }

    @Suppress("Unused")
    fun setProgress(progress: Float)
    {
        properties.progressValue = progress
        properties.calculateContainerValue()
        val sss = properties.progressFilledContainers.subtract(BigDecimal(properties.progressFilledContainers.toInt())).toFloat()

        setCompletedSegments(properties.progressFilledContainers.toInt(), sss)
    }

    /**
     * Callback that will be triggered when a segment is filled.
     */
    @Suppress("Unused")
    fun setCompletedSegmentListener(listener: CompletedSegmentListener)
    {
        this.segmentCompletedListener = listener
    }

    /**
     * Start filling the next unfilled segment.
     * @param timeInMilliseconds total time that will take for the segment to fill.
     */
    fun playSegment(timeInMilliseconds: Long)
    {
        if (properties.segmentedContainerColors[lastCompletedSegment] != null)
            fillRectanglePaint = buildFillRectanglePaint(properties.segmentedContainerColors[lastCompletedSegment]!!)
        else
            fillRectanglePaint = buildFillRectanglePaint(properties.fillColor)

        if (!drawingTimer.isRunning)
        {
            drawingTimer.start(timeInMilliseconds)
        }
    }

    /**
     * Equivalent to [playSegment], but without animation.
     */
    fun incrementCompletedSegments()
    {
        if(currentPartialCompletedSegment!=0f)
        {
            currentPartialCompletedSegment=0f
        }

        if (lastCompletedSegment <= properties.segmentCount)
        {
            currentSegmentProgressInPx = 0
            drawingTimer.reset()
            lastCompletedSegment++
            invalidate()
            segmentCompletedListener?.onSegmentCompleted(lastCompletedSegment)
        }
    }

    /**
     * If segment filling playing, pauses the progress.
     */
    fun pause() = drawingTimer.pause()

    /**
     * If segment filling paused, resumes the progress.
     */
    fun resume() = drawingTimer.resume()

    /**
     * Resets the current bar state, clearing all the segments.
     */
    fun reset() = setCompletedSegments(0)

    /**
     * Checks if the current progress bar filling is paused.
     */
    fun isPaused() = drawingTimer.isPaused

    /**
     * Directly the given number of [completedSegments].
     */
    fun setCompletedSegments(completedSegments: Int, currentPartialCompletedSegment: Float = 0f)
    {
        if (completedSegments <= properties.segmentCount)
        {
            currentSegmentProgressInPx = 0
            this.currentPartialCompletedSegment = currentPartialCompletedSegment
            drawingTimer.reset()
            lastCompletedSegment = completedSegments
            invalidate()
        }
    }

    fun setCompletedSegments(completedSegments: Int)
    {
        setCompletedSegments(completedSegments, 0f)
    }

    /*
      PRIVATE METHODS
       */
    private fun drawContainerRectangles(canvas: Canvas)
    {
        val segmentWidth = segmentWidth

        var leftX = 0
        var rightX = leftX + segmentWidth
        val topY = 0
        val botY = height

        for (i in 0 until properties.segmentCount)
        {
            drawRoundedRect(
                canvas, leftX.toFloat(), topY.toFloat(), rightX.toFloat(), botY.toFloat(),
                containerRectanglePaint)
            leftX += segmentWidth + properties.segmentGapWidth
            rightX = leftX + segmentWidth
        }
    }

    private fun drawCompletedRectangles(canvas: Canvas)
    {
        val segmentWidth = segmentWidth

        var leftX = 0
        var rightX = leftX + segmentWidth
        val topY = 0
        val botY = height

        for (i in 0 until lastCompletedSegment)
        {


            if (containerColor[i] != null)
                drawRoundedRect(
                    canvas, leftX.toFloat(), topY.toFloat(), rightX.toFloat(), botY.toFloat(), buildFillRectanglePaint(containerColor[i]!!))
            else
                drawRoundedRect(canvas, leftX.toFloat(), topY.toFloat(), rightX.toFloat(), botY.toFloat(), fillRectanglePaint)
            leftX += segmentWidth + properties.segmentGapWidth
            rightX = leftX + segmentWidth
        }
    }

    private fun drawCurrentRectangle(canvas: Canvas)
    {
        val segmentWidth = segmentWidth

        val leftX = lastCompletedSegment * (segmentWidth + properties.segmentGapWidth)
        if (currentPartialCompletedSegment != 0f)
        {
            currentSegmentProgressInPx = (segmentWidth * currentPartialCompletedSegment).toInt()
        }

        val rightX = leftX + currentSegmentProgressInPx
        val topY = 0
        val botY = height

        drawRoundedRect(canvas, leftX.toFloat(), topY.toFloat(), rightX.toFloat(), botY.toFloat(), fillRectanglePaint)
    }

    private fun drawRoundedRect(canvas: Canvas, left: Float, top: Float, right: Float, bottom: Float, paint: Paint)
    {

        val path = Path()
        var rx = properties.cornerRadius.toFloat()
        if (rx < 0) rx = 0f
        var ry = 6f
        if (ry < 0) ry = 0f
        val width = right - left
        val height = bottom - top
        if (rx > width / 2) rx = width / 2
        if (ry > height / 2) ry = height / 2
        val widthMinusCorners = width - 2 * rx
        val heightMinusCorners = height - 2 * ry

        with(path) {
            moveTo(right, top + ry)
            rQuadTo(0f, -ry, -rx, -ry) // top-right corner
            rLineTo(-widthMinusCorners, 0f)
            rQuadTo(-rx, 0f, -rx, ry) // top-left corner
            rLineTo(0f, heightMinusCorners)

            rQuadTo(0f, ry, rx, ry) // bottom-left corner
            rLineTo(widthMinusCorners, 0f)
            rQuadTo(rx, 0f, rx, -ry) // bottom-right corner

            rLineTo(0f, -heightMinusCorners)

            close()
        }

        canvas.drawPath(path, paint)
    }

    private fun buildFillRectanglePaint(@ColorInt color: Int): Paint
    {
        val paint = Paint()
        paint.color = color
        paint.style = Paint.Style.FILL
        return paint
    }

    private fun buildContainerRectanglePaint(@ColorInt color: Int): Paint
    {
        val paint = Paint()
        paint.color = color
        paint.style = Paint.Style.FILL
        return paint
    }

    private val segmentWidth: Int
        get() = width / properties.segmentCount - properties.segmentGapWidth
}

interface CompletedSegmentListener
{
    fun onSegmentCompleted(segmentCount: Int)
}
